server/src/main/java/io/crate/execution/ddl/tables/AddColumnRequest.java

```diff
@@ -0,0 +1,286 @@
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.execution.ddl.tables;
+
+import io.crate.analyze.AnalyzedColumnDefinition;
+import io.crate.metadata.GeoReference;
+import io.crate.metadata.IndexType;
+import io.crate.metadata.Reference;
+import io.crate.metadata.RelationName;
+import io.crate.sql.tree.CheckColumnConstraint;
+import io.crate.sql.tree.ColumnPolicy;
+import io.crate.types.ArrayType;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import org.elasticsearch.action.support.master.AcknowledgedRequest;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class AddColumnRequest extends AcknowledgedRequest<AddColumnRequest> {
+
+    private final RelationName relationName;
+
+    // Used only with ADD COLUMN.
+    // Not included into StreamableColumnInfo to avoid adding empty list in most cases.
+    // Checks are merged into mapping without knowing column name as it's contained in expression itself
+    // and thus we can keep them separately which is not the case for other constraints.
+    private final List<StreamableCheckConstraint> checkConstraints = new ArrayList<>();
+
+    private final List<StreamableColumnInfo> columns = new ArrayList<>();
+
+    public AddColumnRequest(StreamInput in) throws IOException {
+        super(in);
+        relationName = new RelationName(in);
+
+        int count = in.readVInt();
+        for (int i = 0; i < count; i++) {
+            checkConstraints.add(StreamableCheckConstraint.readFrom(in));
+        }
+
+        count = in.readVInt();
+        for (int i = 0; i < count; i++) {
+            columns.add(StreamableColumnInfo.readFrom(in, ""));
+        }
+    }
+
+    /**
+     * For dynamic mapping updates.
+     * Fields which cannot be resolved using Reference are initialized with default or null values.
+     */
+    public AddColumnRequest(RelationName relationName,
+                            List<Reference> columnRefs) {
```

Comment:
I'd tend to shift the conversion from `Reference` to `streamableColumnInfo` outside of the request class.
If we're generating a dynamic mapping update we probably don't want to create & allocate a `Reference` instance only to have it converted into a different type moments later.

An option could be to add a `.of(Reference ref)` static factory to `StreamableColumnInfo`. It couples `StreamableColumnInfo` to `Reference` but it's easier to discover & re-use